_labs/lab09.markdown

```diff
@@ -1,6 +1,363 @@
+DSE tracks both runtime values and symbolic constraints, and uses the former to simplify solving the latter during a backtracking search on program computation trees.
+
+We have provided the backbone for a symbolic interpreter using [Z3](https://github.com/Z3Prover/z3).
+You will need to encode a C program into this symbolic interpreter API as well as write the code that drives the dynamic symbolic execution.
+We provide several details on how to do this in the following sections.
+
+This lab assumes that input programs only have integer variables (no pointers or other types of variables) and do not have functions (no `CallInstr`).
+
+#### Understanding Z3
+
+Z3 is a theorem prover developed at Microsoft.
+It's a large and complex tool, so this will serve as a cursory guide for its capabilities and what it can do. Consider a simple, generic system of equations such as the following, where X and Y are integers:
+
+```
+X < Y
+X > 2
+```
+
+Although this example is trivial, think about how you might solve this using any programming language of your choice.
+You may resort to using loops to check numbers or finding a library to handle matrix multiplication.
+This is because most of these programming languages are imperatively-directed, meaning there's a sequence of commands needed to solve the problem.
+
+On the other hand, Z3 has a declarative interface, which in this case means all you need to give it is the list of constraints (in this case, `X < Y` and `X > 2`).
+Plug the following into an [online Z3 solver](https://rise4fun.com/Z3/tutorial/guide) to see the results:
+
+
+```
+(declare-const x Int)
+(declare-const y Int)
+(assert (< x y))
+(assert (> x 2))
+(check-sat)
+(get-model)
+```
+
+Z3 may not give you all possible results that match the constraints but importantly, it verifies satisfiability, which is the key factor that will be leveraged for this DSE engine.
+
+If you're curious about Z3 and want more information, you can check out the following resources:
+
+-   [https://github.com/Z3Prover/z3/wiki/Documentation](https://github.com/Z3Prover/z3/wiki/Documentation)
+-   [https://github.com/Z3Prover/z3/blob/master/examples/c%2B%2B/example.cpp](https://github.com/Z3Prover/z3/blob/master/examples/c%2B%2B/example.cpp)
+-   [https://theory.stanford.edu/\~nikolaj/programmingz3.html](https://theory.stanford.edu/~nikolaj/programmingz3.html)
+
+### Part 1: LLVM Instrumentation
+
+The first component of this dynamic symbolic execution implementation is instrumentation of the input program, which is done in `src/Instrument.cpp`.
+This follows the familiar format and pattern seen in prior labs, except now this LLVM pass will inject various functions defined in `src/Runtime.cpp`, accompanied with the appropriate metadata from each valid LLVM Instruction.
+This will enable DSE to interact with Z3 at runtime.
+Specifically, these are the functions that will require instrumentation (from `include/Instrument.h`):
```

Comment:
Rename all occurrances of `include/Instrument.h` to `include/DSEInstrument.h` similarly for `src/Instrument.cpp`